src/Compilers/Core/Portable/EmbeddedText.cs

```diff
@@ -0,0 +1,345 @@
+        /// <remarks>Reads from the beginning of the stream. Leaves the stream open.</remarks>
+        public static EmbeddedText FromStream(string filePath, Stream stream, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1)
+        {
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            if (filePath.Length == 0)
+            {
+                throw new ArgumentException(CodeAnalysisResources.ArgumentCannotBeEmpty, nameof(filePath));
+            }
+
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            if (!stream.CanRead || !stream.CanSeek)
+            {
+                throw new ArgumentException(CodeAnalysisResources.StreamMustSupportReadAndSeek, nameof(stream));
+            }
+
+            if (stream.Length > int.MaxValue)
+            {
+                throw new IOException(CodeAnalysisResources.StreamIsTooLong);
+            }
+
+            SourceText.ValidateChecksumAlgorithm(checksumAlgorithm);
+
+            return new EmbeddedText(
+                filePath,
+                SourceText.CalculateChecksum(stream, checksumAlgorithm),
+                checksumAlgorithm,
+                CreateBlob(stream));
+        }
+
+        public static EmbeddedText FromBytes(string filePath, byte[] buffer, int length, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1)
+        {
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            if (filePath.Length == 0)
+            {
+                throw new ArgumentException(CodeAnalysisResources.ArgumentCannotBeEmpty, nameof(filePath));
+            }
+
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (length < 0 || length > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            SourceText.ValidateChecksumAlgorithm(checksumAlgorithm);
+
+            return new EmbeddedText(
+                filePath, 
+                SourceText.CalculateChecksum(buffer, 0, length, checksumAlgorithm),
+                checksumAlgorithm, 
+                CreateBlob(buffer, length));
+        }
+
+        private EmbeddedText(string filePath, ImmutableArray<byte> checksum, SourceHashAlgorithm checksumAlgorithm, ImmutableArray<byte> blob)
+        {
+            Debug.Assert(filePath?.Length > 0);
+            Debug.Assert(Cci.DebugSourceDocument.IsSupportedAlgorithm(checksumAlgorithm));
+            Debug.Assert(!blob.IsDefault && blob.Length >= sizeof(int));
+
+            FilePath = filePath;
+            Checksum = checksum;
+            ChecksumAlgorithm = checksumAlgorithm;
+            Blob = blob;
+        }
```

Comment:
I'd move this ctor all the way to the top. We generally declare ctors before any other methods